### PR TITLE
New scheduling code, WeatherNOAA fixes, and documentation updates for DOORBIRD.pm and HARMON.pm.

### PR DESCRIPTION
**Pull request by waynieack**
Added SCHEDULE module for scheduling object state changes via the web UI. This is the backend code for Howard’s ia7 web code. There are 2 sub modules SCHEDULE_Temp and SCHEDULE_Generic.
SCHEDULE_Temp is for thermostat scheduling.
• It allows users to control any thermostat connected to Misterhouse.
• It has objects for setting the thermostat temperatures for each schedule in the MH web UI.
• It has built in occupancy tracking for changing the thermostat temperature settings based on the time and if someone is home or not, including a vacation mode.
• It has a built in outdoor temperature/forecast tracking to change the thermostat temperature settings based on the time and the forecasted temperature or the current outdoor temperature.

SCHEDULE_Generic is for scheduling object state changes for objects that don’t inherit the scheduling code from the Generic_Item.

Generic_Item now has the scheduling code as well so all Generic_Item objects or objects that inherit Generic_Item will be able to have state changes scheduled via Howard's scheduling web UI code. The schedule check code is not added to the main loop until the user adds a schedule via the web UI and the code is removed from the main loop once all the schedule entries are removed.

get_weather has the fixes from Rick Steeves.
WeatherNOAA has fixes from Rick and I.

DOORBIRD.pm and HARMON.pm now have documentation that converts to HTML properly and can be viewed in the local Misterhouse POD documentation page.

DOORBIRD.pm has been updated to post back to Misterhouse using the no_response option.

_Wayne

### CODE REVIEW

File: bin/get_weather [hunk 1]
```diff
@@ -67,99 +60,102 @@ Usage:
...
+#    system("\\mh\\bin\\house.bat show internet weather data");
+#    system("\\mh\\bin\\speak hi");
+#    sleep 3;
 }
 
 if ($return_flag) {
-
-    # Dang ... a 'do pgm' can only return a scalar, not a list :(
-    if ( $data{conditions} ) {
+                                # Dang ... a 'do pgm' can only return a scalar, not a list :(
```
Comment (marcmerlin) [inline bin/get_weather]: formatting broken

File: bin/get_weather [hunk 2]
```diff
@@ -67,99 +60,102 @@ Usage:
...
     }
     else {
         return $forecast;
     }
 }
 else {
-    if ( $data{conditions} ) {
-        print "\nCurrent conditions: $conditions\n";
+    if ($data{conditions}) {
+        #print "\nCurrent conditions: $conditions\n";
```
Comment (marcmerlin) [inline bin/get_weather]: why is this commented out?

File: bin/get_weather [hunk 3]
```diff
@@ -67,99 +60,102 @@ Usage:
...
-    print "Getting the forecast for $parms{zone}, $parms{state}\n";
-    $forecast =
-      print_forecast( $parms{zone}, $parms{state}, undef, undef, undef, 1 );
-    $forecast =~ s/Geo::WeatherNOAA.pm .+\n//;    # Drop geo version
-                                                  #$forecast =~ s/\.\.\./\. /g;
+if ($data{forecast}) {
+    # RICK 160817
+    #print "Getting the forecast for $parms{zone}, $parms{state}\n";
+    print "Getting the forecast for $parms{city}, $parms{state}, $parms{zone}\n";
+    #$forecast = print_forecast($parms{zone}, $parms{state},undef,undef,undef,1);
```
Comment (marcmerlin) [inline bin/get_weather]: please delete commented code

File: bin/get_weather [hunk 4]
```diff
@@ -67,99 +60,102 @@ Usage:
...
 }
 
-if ( $data{forecast} ) {
-    print "Getting the forecast for $parms{zone}, $parms{state}\n";
-    $forecast =
-      print_forecast( $parms{zone}, $parms{state}, undef, undef, undef, 1 );
-    $forecast =~ s/Geo::WeatherNOAA.pm .+\n//;    # Drop geo version
-                                                  #$forecast =~ s/\.\.\./\. /g;
+if ($data{forecast}) {
+    # RICK 160817
```
Comment (marcmerlin) [inline bin/get_weather]: please delete

File: bin/get_weather [hunk 5]
```diff
@@ -67,99 +60,102 @@ Usage:
...
 
 $Geo::WeatherNOAA::proxy_from_env = 1;
 
-if ( $data{conditions} ) {
-    print "\nGetting the current weather for $parms{city}, $parms{state}\n";
-    $conditions =
-      print_current( $parms{city}, $parms{state}, undef, undef, undef, 1 );
+if ($data{conditions}) {
+    # Rick 160816
+    #print "\nGetting the current weather for $parms{city}, $parms{state}\n";
```
Comment (marcmerlin) [inline bin/get_weather]: please delete commented lines

File: bin/get_weather [hunk 6]
```diff
@@ -67,99 +60,102 @@ Usage:
...
+  }
 
-my ( $conditions, $forecast, %data );
+my ($conditions, $forecast, %data);
 my %config_parms;
 
 $parms{city}  = 'Rochester'  unless $parms{city};
-$parms{zone}  = $parms{city} unless $parms{zone};
+#RICK 160817
+#$parms{zone}  = $parms{city} unless $parms{zone};
```
Comment (marcmerlin) [inline bin/get_weather]: please delete commented lines

**Comment by marcmerlin** [comment]
@hollie could you also confirm that the diffs look all bad (lots of unchanged code with simply formatting changes) because @waynieack didn't run a perl reformatter on his changes, or is it another problem?

**Comment by marcmerlin** [comment]
@waynieack I tried to install just your fixes to weather in my tree:
wget --no-check-certificate https://raw.githubusercontent.com/waynieack/misterhouse/9cedc478cf1ca56f2db4ab898411f97be920395f/bin/get_weather
wget --no-check-certificate https://raw.githubusercontent.com/waynieack/misterhouse/9cedc478cf1ca56f2db4ab898411f97be920395f/lib/site/Geo/WeatherNOAA.pm

but I still don't seem to be getting weather with this, am I supposed to change the way I call it?
gargamel:/var/local/src/misterhouse# mh/bin/get_weather -zone "SAN JOSE" -city "SAN JOSE" -state CA

Getting the current weather for SAN JOSE, CA, SAN JOSE
Getting the forecast for SAN JOSE, CA, SAN JOSE
gargamel:/var/local/src/misterhouse# cat /var/local/src/misterhouse/data/web/weather_forecast.txt
As of :
gargamel:/var/local/src/misterhouse#

**Comment by hollie** [comment]
@waynieack first of all thanks for this pull request!

As @marcmerlin mentions: normally it is indeed the idea to split up different fixes or adaptations into different pull requests because this makes the code easier to review. You can do this by making a new branch in your working copy, as described here: https://github.com/hollie/misterhouse/wiki/Contributing#developing-using-git

Then a general remark: you don't have to comment out lines and add your name in front. Since we're using git we know exactly who is responsible for a specific code change ;-)

There are indeed a few changes that are only related to whitespaces that are changed. Did you run a specific formatter on the code to do that? We have a specific set of Perltidy settings that we use to format the code in this project. For more info please see here: https://github.com/hollie/misterhouse/wiki/Contributing#using-perltidy-to-make-consistent-formatted-code

I think that there is lots of interesting code in this pull request, so if you could update the changes requested by Marc and rerun the formatting on your code then we'll look into integrating it into master.

Note: if you update your branch this pull request should also receive the updates. This is one of the reasons you should normally not create a pull request from your master branch.

**Comment by waynieack** [comment]
Sorry, I'm new to git. I had some issues with conflicts and made a mess
with the reverts. For get_weather, Steve worked on that and emailed it to
me, I'll clean it up, (figure out how to make a new branch:)) make a branch
and create a new pull request.


The changes to WeatherNOAA.pm were all related to this issue, is that OK to
do in a single branch?


Do you want me to cancel the current pull request?


_Wayne

**Comment by marcmerlin** [comment]
anwers
1) you don't need one pull request per file, but per logical change. So in your case, you should put both weather files in the same pull request (also each pull request should create a tree that is consistent and works after installing just that pull request)
2) in this case, you may want to take a clean tree on your side, make a branch, copy your weather changes in it, make a pull request for just that.
Then do another tree with another branch with a different name and put your unrelated pull request with related new features in that one.
After you upload those 2 and they look good online, you can delete this one.
3) remember to run perltidy so that you don't send diffs that undo the current white space rules and make no other code changes

**Comment by waynieack** [comment]
The zone is a 3 letter code from:
http://forecast.weather.gov/product_sites.php?site=CRH&product=ZFP.

The easiest way I've found to get the zone is to search your zip or
city/state in the "Local forecast by" search box at the top left, then get
the 3 letter zone from the URL you are redirected to. (the zone is after
"&site=" in the URL)

If I search Baton Rouge, LA I get redirected to
http://forecast.weather.gov/MapClick.php?CityName=Baton+Rouge&state=LA&site=LIX&lat=30.449&lon=-91.126#.WAgBXiSPFv0

So my zone is: LIX

San Jose is: MTR

Call it like this:
get_weather -zone MTR -city "SAN JOSE" -state CA


It looks like I can hack together a way to get the zone with the city and
state using a post to the search box and pulling the zone from the site=
field of the location header. The only issue with this is that if any
changes are made to the page it will break, so I don't know if its worth
it. What do you think?

If we go this route then get_weather and the user mh.private.ini would not
be changed and I could just update WeatherNOAA.pm.


_Wayne

**Comment by waynieack** [comment]
Just to be sure I don't make more of a mess than I already have :)

I set up the hook with perl tidy so all my commits will run through it now.

Copy my changed files out.
Reset all the changes from the last commit
git reset --hard b670e1579d9191a38a04400ddb8c7ba87bb434ba
git commit -am "Revert to move to changes to branches"

git checkout -b Schedule

copy SCHEDULE.pm and Generic_Item.pm back

git stage SCHEDULE.pm
git stage Generic_Item.pm
git commit -m "Completed SCHEDULE.pm beta, adding more options, some bug
fixes, and Documentation. Added scheduling code in Generic_Item, so all
Generic_Item objects or objects that inherit Generic_Item will have
scheduling options in the UI."
git push origin Schedule



git checkout -b Harmon

copy HARMON.pm back

git stage HARMON.pm
git commit -m "Updated Documentation"
git push origin Harmon




git checkout -b Doorbird

copy DOORBIRD.pm back

git stage DOORBIRD.pm
git commit -m "Updated Documentation"
git push origin Doorbird




Fix issues with get_weather noted in current pull request.

git checkout -b WeatherNOAA

copy WeatherNOAA.pm and get_weather back

git stage WeatherNOAA.pm
git stage get_weather
git commit -m "Fixes for broken code due to NOAA site changes. Bug #631"
git push origin WeatherNOAA


cancel current pull request
create a pull request for each of the above.


Let me know if I am on the right track.

_Wayne

**Comment by marcmerlin** [comment]
First, I missed the part that I have to give new arguments to get_weather, it works now, thank you.
As for how to do the changes, looking at what you're saying, I think it'll work, but ultimately you'll see what it looks like in your github tree when you upload.
You can see if that diff looks clean and self contained before you send a pull request to this tree. Github makes it easy to see what you're sending before you send it.

**Comment by waynieack** [comment]
I made 3 new pull requests, the Harmon and Doorbird requests failed some of
the checks but I am not sure why. The code did not change and the checks
passes on the first pull request.

I did however run all the code through perl tidy with the hook, so there
are many changes in the diffs due to that.

I updated the weather NOAA code so the zone is not needed anymore.


   -
   Fixes for broken code due to NOAA site changes. Bug #631
   <https://github.com/hollie/misterhouse/pull/635>
   <https://github.com/hollie/misterhouse/pull/635#partial-pull-merging>
   #635
   -
   Doorbird documentation updates.
   <https://github.com/hollie/misterhouse/pull/634>
   <https://github.com/hollie/misterhouse/pull/634#partial-pull-merging>
   #634
   -
   Harmon documentation updates
   <https://github.com/hollie/misterhouse/pull/633>
   <https://github.com/hollie/misterhouse/pull/633#partial-pull-merging>
   #633


_Wayne

**Comment by hollie** [comment]
Hello Wayne,

regarding the tests, yes I'm aware of the problem and I'm looking into it. Might take until the weekend before I find some time to clear that out.

Thanks for splitting the contributions in separate pull requests.

Best regards,
 Lieven.

**Pull request closed without merging**